res/controllers/Numark-Mixtrack-Platinum-scripts.js

```diff
@@ -0,0 +1,1415 @@
+            loaded: 0x00,
+            playing: 0x7F,
+        });
+    }
+    this.hotcues = this.hotcue_buttons;
+
+    this.pitch = new components.Pot({
+        inKey: 'rate',
+        invert: true,
+    });
+    if (!this.active) {
+        this.pitch.firstValueReceived = true;
+    }
+
+    var pitch_or_keylock = function (channel, control, value, status, group) {
+        if (this.other.inGetValue() > 0.0 && this.isPress(channel, control, value, status)) {
+            // toggle keylock, both keys pressed
+            engine.setValue(this.group, "keylock", !engine.getValue(this.group, "keylock"));
+        }
+        else {
+            components.Button.prototype.input.call(this, channel, control, value, status, group);
+        }
+    };
+    this.pitch_bend_up = new components.Button({
+        inKey: 'rate_temp_up',
+        input: pitch_or_keylock,
+    });
+    this.pitch_bend_down = new components.Button({
+        inKey: 'rate_temp_down',
+        input: pitch_or_keylock,
+    });
+    this.pitch_bend_up.other = this.pitch_bend_down;
+    this.pitch_bend_down.other = this.pitch_bend_up;
+
+    var key_up_or_down = function (channel, control, value, status, group) {
+        this.is_pressed = this.isPress(channel, control, value, status);
+        if (this.is_pressed) {
+            if (this.other.is_pressed) {
+                // reset if both buttons are pressed
+                this.inSetValue(0.0);
+            }
+            else {
+                var old = this.inGetValue();
+                // the gui only allows 7 steps, we replicate that behavior here
+                if (Math.abs(old + this.direction) < 8.0) {
+                    this.inSetValue(Math.round(old + 1.0 * this.direction));
+                }
+            }
+        }
+    };
+    this.key_up = new components.Button({
+        inKey: 'pitch_adjust',
+        direction: 1,
+        input: key_up_or_down,
+    });
+    this.key_down = new components.Button({
+        inKey: 'pitch_adjust',
+        direction: -1,
+        input: key_up_or_down,
+    });
+    this.key_up.other = this.key_down;
+    this.key_down.other = this.key_up;
+
+    loop_base = function(midino, obj) {
+        return _.assign({
+            midi: [0x94 + midi_chan, midino],
+            on: 0x01,
+            sendShifted: true,
+            shiftChannel: true,
+            shiftOffset: -0x10,
+        }, obj);
+    };
+
+    this.alternate_manloop = new components.ComponentContainer({
+        loop_in: new components.HotcueButton(loop_base(0x38, {
+            number: 5,
+        })),
+        loop_out: new components.HotcueButton(loop_base(0x39, {
+            number: 6,
+        })),
+        loop_toggle: new components.HotcueButton(loop_base(0x32, {
+            number: 7,
+        })),
+        loop_halve: new components.HotcueButton(loop_base(0x34, {
+            number: 8,
+        })),
+        loop_double: new components.HotcueButton(loop_base(0x35, {
+            number: 8,
```

Comment:
There are two "hotcue button 8"s here. Did you accidentally do an extra copy & paste?

Reply:
I'm confused. That's what `sendShifted` is for and it's only hotcue 8 that has duplicate buttons here.

Reply:
Okay, please add a comment explaining this.